### ISSUE TITLE
Chainable baseURL

### ISSUE THREAD

**Issue by idiotWu** [service-mocker/service-mocker#22]
Currently(v1.0.x), `router.base()` will always resolve to the **origin** of current page. For example:

```js
const { router } = createServer();

router.base('/api'); // => http://localhost:3000/api
```

However, this will be counterintuitive when we code as:

```js
const { router } = createServer('/api');

router.base('/v1'); // => http://localhost:3000/v1
```

Instead of getting a path of `/api/v1`, we got `/v1` :(. So I'm wondering if we should resolve the given `baseURL` against current `baseURL`, just like the mount-paths in express apps.

CC @VincentBel

**Comment by idiotWu** [comment]
Some potential issues:

### 1. Syntax

Should relative paths be prefixed with `'.'`?

```js
const { router } = createServer('/api');

router.base('/v1'); 
// OR
router.base('./v1');
```

The native `URL` accepts both relative and absolute paths (notice the trailing slash):

```js
new URL('/v1', 'http://localhost:3000/api'); // => http://localhost:3000/v1
new URL('./v1', 'http://localhost:3000/api'); // => http://localhost:3000/v1
new URL('./v1', 'http://localhost:3000/api/'); // => http://localhost:3000/api/v1
```

**Comment by VincentBel** [comment]
@idiotWu Summary of some of our discussions:

1. Servers with different origin should use `createServer`, rather than `router.base`(because they are different servers, `router.base` is counter-intuitive)

 ```js
// Good
const apiRouter = createServer('/api').router
const githubApiRouter = createServer('https://api.github.com').router

// Bad
const server = createServer('/api')
const apiRouter = server.router
const githubApiRouter = server.router.base('/https://api.github.com')

// Also bad
const server = createServer()
const apiRouter = server.router.base('/api')
const githubApiRouter = server.router.base('/https://api.github.com')
```

2. Chainable baseURL can also be used to split routes. Example: https://gist.github.com/VincentBel/6a7d35e6a2999e2aa90d065a77831bfd


Feel free to point out faults if exists.

**Comment by idiotWu** [comment]
@VincentBel so the given `baseURL` should be regarded as a kind of "prefixes" for the path?

**Comment by VincentBel** [comment]
@idiotWu yeah, I think we don't provide `router.base` as a way to create a new server in purpose. So I also think it is more a bug fix.

**Comment by idiotWu** [comment]
@VincentBel one more thing, as we've discussed, the `router.base()` method is designed for "prefixes", or we can say **relative paths only**. So should we reject all absolute paths or just throw a warning?

```js
router.base('http://a.com'); // die or show warning?
```

**Comment by VincentBel** [comment]
@idiotWu I think `die` is better because we don't mean to support this approach.😑

**Comment by VincentBel** [comment]
and warnings are just ignored by developers. 😅

**Comment by idiotWu** [comment]
Resolved in #23 #25
